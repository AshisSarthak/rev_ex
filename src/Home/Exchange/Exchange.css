.exchange-from-container {
  display: flex;
  width: 80%;
  justify-content: space-around;
  align-items: center;
  margin-top: -10px;
  margin-left: 10%;
  height: 200px;
  border: 1px solid #10cbf7;
  border-radius: 10px;
}

.exchange-from-select {
  font-size: 33px;
  border: 1px solid transparent;
  border-radius: 10px;
  padding: 10px;
  outline: none;
}
.exchange-from-input {
  font-size: 33px;
  padding: 10px;
  border-radius: 4px;
  border: 1px solid transparent;
  outline: none;
  margin-left: 10px;
  width: 220px;
}

.exchange-rate {
  position: relative;
  bottom: -10px;
  background-color: #10cbf7;
  width: 20%;
  left: calc((100% - 20%) / 2);
  padding: 5px;
  text-align: center;
  border-radius: 4px;
  font-size: 20px;
}

.exchange-to-label {
  font-size: 4em;
  font-weight: 100;
  color: white;
}

.convert-button {
  float: right;
  padding: 10px;
  border: 1px solid transparent;
  border-radius: 4px;
  cursor: pointer;
  margin: 20px 200px 0px 0px;
  font-size: 20px;
  background-color: #10cbf7;
}
