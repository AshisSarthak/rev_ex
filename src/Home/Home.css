.home-header {
  background: #333;
  display: flex;
  justify-content: center;
  height: 50px;
  align-items: center;
}

.header-nav {
  width: 100px;
  border: 2px solid #10cbf7;
  color: #10cbf7;
  height: 30px;
  display: flex;
  justify-content: center;
  align-items: center;
  cursor: pointer;
}

.header-nav:first-child {
  border-right: none;
}

.selected-nav {
  background-color: #10cbf7;
  color: #333;
}
