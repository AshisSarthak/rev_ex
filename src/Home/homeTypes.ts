export type RawFX = {
  rates: Map<string, number>;
  base: string;
  date: string;
};

export type TransactionData = {
  id: number;
  timestamp: number;
  balance: number;
};

export type PocketData = {
  balance: number;
  transactions: Array<TransactionData>;
};
