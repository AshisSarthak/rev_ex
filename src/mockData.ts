export const demoPacketData = {
  balance: 100,
  transactions: [
    {
      id: 1,
      timestamp: 1000230,
      balance: 50,
    },
    {
      id: 2,
      timestamp: 1000230,
      balance: 80,
    },
    {
      id: 3,
      timestamp: 1000230,
      balance: 90,
    },
  ],
};

export enum CURRENCY_SYMBOL {
  USD = "$",
  EUR = "€",
  GBP = "£",
}

export enum CURRENCYS {
  USD = "USD",
  EUR = "EUR",
  GBP = "GBP",
}
