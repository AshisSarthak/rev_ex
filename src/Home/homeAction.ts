import { EXCHANGE_API } from "./homeConstants";
import { RawFX } from "./homeTypes";

export const fetchCurrentFXRateAction = (rawFxRates: RawFX) => ({
  type: "FETCH_FX_RATE",
  payload: rawFxRates,
});

export const fetchCurrentFXRate = (baseCurrency: string = "USD") => async (
  dispatch: any
): Promise<void> => {
  const fxRates: RawFX = await fetch(
    `${EXCHANGE_API}${baseCurrency}`
  ).then((resp) => resp.json());
  console.log(fxRates);
  dispatch(fetchCurrentFXRateAction(fxRates));
};
