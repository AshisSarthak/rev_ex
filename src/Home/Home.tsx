import React from "react";
import { connect } from "react-redux";
import "./Home.css";
import Pocket from "./Pockets/Pockets";
import Exchange from "./Exchange/Exchange";
import { CURRENCYS } from "../mockData";

export type HomeProps = {
  gbpPocket: any;
  usdPocket: any;
  eurPocket: any;
};

export type HomeState = {
  currentPage: string;
};

export class Home extends React.Component<HomeProps, HomeState> {
  state = {
    currentPage: "home",
  };

  getHeaderClass = (headerPage: string) => {
    let headerString = "header-nav";
    if (this.state.currentPage === headerPage) {
      headerString += " selected-nav";
    }
    return headerString;
  };

  updateHeaderNav = (headerPage: string) => {
    if (this.state.currentPage !== headerPage) {
      this.setState({
        currentPage: headerPage,
      });
    }
  };

  render() {
    const { gbpPocket, usdPocket, eurPocket } = this.props;
    const { currentPage } = this.state;
    return (
      <div className="home">
        <header className="home-header">
          <nav
            className={this.getHeaderClass("home")}
            onClick={() => this.updateHeaderNav("home")}
          >
            Home
          </nav>
          <nav
            className={this.getHeaderClass("exchange")}
            onClick={() => this.updateHeaderNav("exchange")}
          >
            Exchange
          </nav>
        </header>
        {currentPage === "home" ? (
          <>
            <div className="pockets-header">Your Pockets</div>
            <section className="pockets-details">
              <Pocket pocket={gbpPocket} currency={CURRENCYS.GBP}></Pocket>
              <Pocket pocket={usdPocket} currency={CURRENCYS.USD}></Pocket>
              <Pocket pocket={eurPocket} currency={CURRENCYS.EUR}></Pocket>
            </section>
          </>
        ) : (
          <Exchange />
        )}
      </div>
    );
  }
}

const mapStateToProps = (state: any) => ({
  fxRate: state.fxRates.fxRates,
  gbpPocket: state.fxRates.pockets.GBP,
  usdPocket: state.fxRates.pockets.USD,
  eurPocket: state.fxRates.pockets.EUR,
});

const mapDispatchToProps = {};

export default connect(mapStateToProps, mapDispatchToProps)(Home);
