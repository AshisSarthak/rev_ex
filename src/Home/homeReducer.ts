import { demoPacketData } from "../mockData";

export const defaultState = {
  fxRates: undefined,
  pockets: {
    gbp: { ...demoPacketData },
    usd: { ...demoPacketData },
    eur: { ...demoPacketData },
  },
};

const homeReducer = (state = defaultState, action: any) => {
  switch (action.type) {
    case "FETCH_FX_RATE":
      return {
        ...state,
        fxRates: action.payload,
      };
    default:
      return state;
  }
};

export default homeReducer;
