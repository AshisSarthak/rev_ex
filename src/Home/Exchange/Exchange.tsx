import React from "react";
import { connect } from "react-redux";
import { fetchCurrentFXRate, convertToPocket } from "../homeAction";
import { CURRENCYS } from "../../mockData";
import SelectEl from "../Common/Select";
import InputEl from "../Common/Input";
import "./Exchange.css";
import { getCurrencySymbol } from "../Pockets/Pockets";

const defaultOptionList = [CURRENCYS.USD, CURRENCYS.EUR, CURRENCYS.GBP];
const validInputReg = new RegExp(/^[+-]?((\d+(\.\d{1,2})?)|(\.\d+))$/);

export const getToCurrencyList = (targetCur: CURRENCYS) =>
  defaultOptionList.filter((val: CURRENCYS) => val !== targetCur);

export type ExchangeState = {
  fromCurrency: CURRENCYS;
  fromAmount: number | undefined;
  toCurrencyList: Array<CURRENCYS>;
  toCurrency: CURRENCYS;
  toAmount: number | undefined;
};

export type Props = {
  fetchCurrentFXRate: Function;
  convertToPocket: Function;
  fxRate: any;
};

export class Exchange extends React.Component<Props, ExchangeState> {
  state = {
    fromCurrency: CURRENCYS.USD,
    toCurrencyList: getToCurrencyList(CURRENCYS.USD),
    toCurrency: getToCurrencyList(CURRENCYS.USD)[0],
    toAmount: undefined,
    fromAmount: undefined,
  };

  componentDidMount() {
    this.props.fetchCurrentFXRate(this.state.fromCurrency);
  }

  selectFromChangeHandler = async (value: CURRENCYS) => {
    await this.props.fetchCurrentFXRate(value);
    this.setState({
      fromCurrency: value,
      toCurrencyList: getToCurrencyList(value),
      toCurrency: getToCurrencyList(value)[0],
      toAmount:
        this.state.fromAmount! *
        this.getExchangeVal(getToCurrencyList(value)[0]),
    });
  };

  selectToChangeHandler = (value: any) => {
    this.setState({
      toCurrency: value,
      toAmount: this.state.fromAmount! * this.getExchangeVal(value),
    });
  };

  selectFromHandler = (value: any) => {
    this.setState({
      fromAmount: value,
      toAmount: value * this.getExchangeVal(),
    });
  };

  getExchangeVal = (targetCurrency: CURRENCYS | undefined = undefined) => {
    if (this.props.fxRate && this.props.fxRate.rates) {
      return targetCurrency
        ? this.props.fxRate.rates[targetCurrency!]
        : this.props.fxRate.rates[this.state.toCurrency];
    }
  };

  handleConvertPocket = () => {
    const { fromAmount, fromCurrency, toAmount, toCurrency } = this.state;
    this.props.convertToPocket({
      fromAmount,
      fromCurrency,
      toAmount,
      toCurrency,
    });
  };

  render() {
    const exchangeRateString =
      this.getExchangeVal() &&
      `1${getCurrencySymbol(
        this.state.fromCurrency
      )} = ${this.getExchangeVal()}${getCurrencySymbol(this.state.toCurrency)}`;

    return (
      <section className="exchange-container">
        <div className="exchange-rate">{exchangeRateString}</div>
        <div className="exchange-from-container">
          <div className="exchange-box">
            <SelectEl
              onChange={this.selectFromChangeHandler}
              optionLists={defaultOptionList}
              currentVal={this.state.fromCurrency}
            />
            <InputEl
              onChange={this.selectFromHandler}
              value={this.state.fromAmount}
            />
          </div>
          <div className="exchange-to-label"> to </div>
          <div className="exchange-box">
            <SelectEl
              onChange={this.selectToChangeHandler}
              optionLists={this.state.toCurrencyList}
              currentVal={this.state.toCurrency}
            />
            <InputEl value={this.state.toAmount} disabled />
          </div>
        </div>

        <button
          className="convert-button"
          disabled={!validInputReg.test(this.state.fromAmount!)}
          onClick={this.handleConvertPocket}
        >
          Convert to Pocket
        </button>
      </section>
    );
  }
}

const mapStateToProps = (state: any) => ({
  fxRate: state.fxRates.fxRates,
});

const mapDispatchToProps = {
  fetchCurrentFXRate,
  convertToPocket,
};

export default connect(mapStateToProps, mapDispatchToProps)(Exchange);
