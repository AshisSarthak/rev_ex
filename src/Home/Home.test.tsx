import React from "react";
import { shallow, mount } from "enzyme";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import { Home } from "./Home";
import { demoPacketData } from "../mockData";
import Pocket from "./Pockets/Pockets";
import { Exchange } from "./Exchange/Exchange";
import { defaultState } from "./homeReducer";

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
let store;
describe("Home Element", () => {
  beforeAll(() => {
    store = mockStore({ ...defaultState });
    store.clearActions();
  });

  it("Loads the default Dropdown", () => {
    const wrapper = shallow(
      <Home
        gbpPocket={demoPacketData}
        usdPocket={demoPacketData}
        eurPocket={demoPacketData}
      />
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.state("currentPage")).toBe("home");
    const pockets = wrapper.find(Pocket);
    expect(pockets).toHaveLength(3);

    expect(wrapper.find(".pockets-header")).toHaveLength(1);
    expect(wrapper.find(Exchange)).toHaveLength(0);
  });

  it("Change the nav options", async () => {
    const wrapper = shallow(
      <Home
        gbpPocket={demoPacketData}
        usdPocket={demoPacketData}
        eurPocket={demoPacketData}
      />
    );
    const navs = wrapper.find("nav");
    expect(navs).toHaveLength(2);
    navs.at(1).simulate("click");
    await wrapper.update();
    expect(wrapper.state("currentPage")).toBe("exchange");

    expect(wrapper.find(".pockets-header")).toHaveLength(0);
    expect(wrapper.find(Exchange)).toBeTruthy();

    navs.at(0).simulate("click");
    await wrapper.update();
    expect(wrapper.state("currentPage")).toBe("home");

    expect(wrapper.find(".pockets-header")).toHaveLength(1);
    expect(wrapper.find(Exchange)).toHaveLength(0);

    navs.at(0).simulate("click");
    await wrapper.update();
    expect(wrapper.state("currentPage")).toBe("home");
  });

  it("Loads the redux connect", () => {
    const wrapper = mount(
      <Provider store={store}>
        <Home
          gbpPocket={demoPacketData}
          usdPocket={demoPacketData}
          eurPocket={demoPacketData}
        />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
