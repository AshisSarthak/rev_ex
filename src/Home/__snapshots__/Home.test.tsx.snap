// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Element Loads the default Dropdown 1`] = `
<div
  className="home"
>
  <header
    className="home-header"
  >
    <nav
      className="header-nav selected-nav"
      onClick={[Function]}
    >
      Home
    </nav>
    <nav
      className="header-nav"
      onClick={[Function]}
    >
      Exchange
    </nav>
  </header>
  <div
    className="pockets-header"
  >
    Your Pockets
  </div>
  <section
    className="pockets-details"
  >
    <Pocket
      currency="GBP"
      pocket={
        Object {
          "balance": 100,
          "transactions": Array [
            Object {
              "balance": 50,
              "id": 1,
              "timestamp": 1000230,
            },
            Object {
              "balance": 80,
              "id": 2,
              "timestamp": 1000230,
            },
            Object {
              "balance": 90,
              "id": 3,
              "timestamp": 1000230,
            },
          ],
        }
      }
    />
    <Pocket
      currency="USD"
      pocket={
        Object {
          "balance": 100,
          "transactions": Array [
            Object {
              "balance": 50,
              "id": 1,
              "timestamp": 1000230,
            },
            Object {
              "balance": 80,
              "id": 2,
              "timestamp": 1000230,
            },
            Object {
              "balance": 90,
              "id": 3,
              "timestamp": 1000230,
            },
          ],
        }
      }
    />
    <Pocket
      currency="EUR"
      pocket={
        Object {
          "balance": 100,
          "transactions": Array [
            Object {
              "balance": 50,
              "id": 1,
              "timestamp": 1000230,
            },
            Object {
              "balance": 80,
              "id": 2,
              "timestamp": 1000230,
            },
            Object {
              "balance": 90,
              "id": 3,
              "timestamp": 1000230,
            },
          ],
        }
      }
    />
  </section>
</div>
`;

exports[`Home Element Loads the redux connect 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Home
    eurPocket={
      Object {
        "balance": 100,
        "transactions": Array [
          Object {
            "balance": 50,
            "id": 1,
            "timestamp": 1000230,
          },
          Object {
            "balance": 80,
            "id": 2,
            "timestamp": 1000230,
          },
          Object {
            "balance": 90,
            "id": 3,
            "timestamp": 1000230,
          },
        ],
      }
    }
    gbpPocket={
      Object {
        "balance": 100,
        "transactions": Array [
          Object {
            "balance": 50,
            "id": 1,
            "timestamp": 1000230,
          },
          Object {
            "balance": 80,
            "id": 2,
            "timestamp": 1000230,
          },
          Object {
            "balance": 90,
            "id": 3,
            "timestamp": 1000230,
          },
        ],
      }
    }
    usdPocket={
      Object {
        "balance": 100,
        "transactions": Array [
          Object {
            "balance": 50,
            "id": 1,
            "timestamp": 1000230,
          },
          Object {
            "balance": 80,
            "id": 2,
            "timestamp": 1000230,
          },
          Object {
            "balance": 90,
            "id": 3,
            "timestamp": 1000230,
          },
        ],
      }
    }
  >
    <div
      className="home"
    >
      <header
        className="home-header"
      >
        <nav
          className="header-nav selected-nav"
          onClick={[Function]}
        >
          Home
        </nav>
        <nav
          className="header-nav"
          onClick={[Function]}
        >
          Exchange
        </nav>
      </header>
      <div
        className="pockets-header"
      >
        Your Pockets
      </div>
      <section
        className="pockets-details"
      >
        <Pocket
          currency="GBP"
          pocket={
            Object {
              "balance": 100,
              "transactions": Array [
                Object {
                  "balance": 50,
                  "id": 1,
                  "timestamp": 1000230,
                },
                Object {
                  "balance": 80,
                  "id": 2,
                  "timestamp": 1000230,
                },
                Object {
                  "balance": 90,
                  "id": 3,
                  "timestamp": 1000230,
                },
              ],
            }
          }
        >
          <section
            className="pocket-container"
          >
            <div
              className="pocket-currency"
            >
              <div
                className="currency-balance"
              >
                100.00
                <div
                  className="currency-symbol"
                >
                  £
                </div>
              </div>
            </div>
            <div
              className="exchange-table"
            >
              <div
                className="exchange-entries"
              >
                <div
                  className="exchange-header"
                >
                  Timestamp
                </div>
                <div
                  className="exchange-header"
                >
                  Balance
                </div>
              </div>
               
              <div
                className="exchange-entries"
                key="0"
              >
                <div>
                  1000230
                </div>
                <div>
                  50
                  £
                </div>
              </div>
              <div
                className="exchange-entries"
                key="1"
              >
                <div>
                  1000230
                </div>
                <div>
                  80
                  £
                </div>
              </div>
              <div
                className="exchange-entries"
                key="2"
              >
                <div>
                  1000230
                </div>
                <div>
                  90
                  £
                </div>
              </div>
            </div>
          </section>
        </Pocket>
        <Pocket
          currency="USD"
          pocket={
            Object {
              "balance": 100,
              "transactions": Array [
                Object {
                  "balance": 50,
                  "id": 1,
                  "timestamp": 1000230,
                },
                Object {
                  "balance": 80,
                  "id": 2,
                  "timestamp": 1000230,
                },
                Object {
                  "balance": 90,
                  "id": 3,
                  "timestamp": 1000230,
                },
              ],
            }
          }
        >
          <section
            className="pocket-container"
          >
            <div
              className="pocket-currency"
            >
              <div
                className="currency-balance"
              >
                100.00
                <div
                  className="currency-symbol"
                >
                  $
                </div>
              </div>
            </div>
            <div
              className="exchange-table"
            >
              <div
                className="exchange-entries"
              >
                <div
                  className="exchange-header"
                >
                  Timestamp
                </div>
                <div
                  className="exchange-header"
                >
                  Balance
                </div>
              </div>
               
              <div
                className="exchange-entries"
                key="0"
              >
                <div>
                  1000230
                </div>
                <div>
                  50
                  $
                </div>
              </div>
              <div
                className="exchange-entries"
                key="1"
              >
                <div>
                  1000230
                </div>
                <div>
                  80
                  $
                </div>
              </div>
              <div
                className="exchange-entries"
                key="2"
              >
                <div>
                  1000230
                </div>
                <div>
                  90
                  $
                </div>
              </div>
            </div>
          </section>
        </Pocket>
        <Pocket
          currency="EUR"
          pocket={
            Object {
              "balance": 100,
              "transactions": Array [
                Object {
                  "balance": 50,
                  "id": 1,
                  "timestamp": 1000230,
                },
                Object {
                  "balance": 80,
                  "id": 2,
                  "timestamp": 1000230,
                },
                Object {
                  "balance": 90,
                  "id": 3,
                  "timestamp": 1000230,
                },
              ],
            }
          }
        >
          <section
            className="pocket-container"
          >
            <div
              className="pocket-currency"
            >
              <div
                className="currency-balance"
              >
                100.00
                <div
                  className="currency-symbol"
                >
                  €
                </div>
              </div>
            </div>
            <div
              className="exchange-table"
            >
              <div
                className="exchange-entries"
              >
                <div
                  className="exchange-header"
                >
                  Timestamp
                </div>
                <div
                  className="exchange-header"
                >
                  Balance
                </div>
              </div>
               
              <div
                className="exchange-entries"
                key="0"
              >
                <div>
                  1000230
                </div>
                <div>
                  50
                  €
                </div>
              </div>
              <div
                className="exchange-entries"
                key="1"
              >
                <div>
                  1000230
                </div>
                <div>
                  80
                  €
                </div>
              </div>
              <div
                className="exchange-entries"
                key="2"
              >
                <div>
                  1000230
                </div>
                <div>
                  90
                  €
                </div>
              </div>
            </div>
          </section>
        </Pocket>
      </section>
    </div>
  </Home>
</Provider>
`;
